src/main/java/com/gargoylesoftware/htmlunit/html/HtmlLabel.java

```diff
@@ -129,20 +129,35 @@ public HtmlElement getReferencedElement() {
         final String elementId = getForAttribute();
         if (!ATTRIBUTE_NOT_DEFINED.equals(elementId)) {
             try {
-                return ((HtmlPage) getPage()).getHtmlElementById(elementId);
+                HtmlElement element = ((HtmlPage) getPage()).getHtmlElementById(elementId);
+                if (isLabelableElement(element)) {
+                	return element;
+                }
             }
             catch (final ElementNotFoundException e) {
                 return null;
             }
-        }
-        for (final DomNode element : getChildren()) {
-            if (element instanceof HtmlInput) {
-                return (HtmlInput) element;
-            }
+        } else {
+	        for (final DomNode element : getChildren()) {
+	            if (isLabelableElement(element)) {
+	                return (HtmlElement) element;
+	            }
+	        }
         }
         return null;
     }
 
+    private boolean isLabelableElement(DomNode element) {
+    	// TODO maybe we should introduce an interface LabelabelElement for this?
```

Comment:
Think you are right here, a marker interface (similar to DisabledElement) looks like the right solution here.
I guess there is a typo in you suggestion - LabelableElement. Maybe you can add some spec link to the interface description.